clippy_lints/src/stacked_ifs.rs

```diff
@@ -0,0 +1,78 @@
+use clippy_utils::diagnostics::span_lint_and_help;
+use clippy_utils::in_constant;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Detects `if` statements where the condition is another `if` statement.
+    /// ### Why is this bad?
+    /// This makes code hard to read.
+    /// ### Example
+    /// ```no_run
+    /// if if some_condition {
+    ///     some_value
+    /// } else {
+    ///     some_other_value
+    /// } == another_value {
+    ///     // Do something.
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let value = if some_condition {
+    ///     some_value
+    /// } else {
+    ///     some_other_value
+    /// };
+    /// if value == another_value {
+    ///     // Do something.
+    /// }
+    /// ```
+    #[clippy::version = "1.79.0"]
+    pub STACKED_IFS,
+    style,
+    "finds if statements with another if statement as the condition"
+}
+
+declare_lint_pass!(StackedIfs => [STACKED_IFS]);
+
+impl LateLintPass<'_> for StackedIfs {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
+        // Ensure that the expression isn't a part of a constant or macro expansion.
+        if expr.span.from_expansion() || in_constant(cx, expr.hir_id) {
```

Comment:
The `in_constant` check can be removed as we do want to lint this inside `const` contexts